File: week1/valid_parenthesis.dart
bool isValid(String s)  {
    Map<String,String> lookupTable = {
      ")":"(",
      "}":"{",
      "]":"["
    };
 
    
    List<String> stack = [];
    
    for(int i=0; i<s.length; i++){
      var l = stack.length;
      
      if(lookupTable.containsKey(s[i])){
        if(l>0 && stack.last == lookupTable[s[i]]){
          stack.removeLast();
        }else{
          return false;
        }
      }else{
        stack.add(s[i]);
      }
      
    }
    
    return stack.isEmpty;
}

File: week1/valid_parenthesis.go
func isValid(s string) bool {
    lookupTable := make(map[rune]rune)
    lookupTable[')']='('
    lookupTable['}']='{'
    lookupTable[']']='['
    
    var stack []rune
    
    for _,char := range s {
        l := len(stack)
        if _,ok := lookupTable[char]; ok  {
            if l>0 && stack[l-1] == lookupTable[char] {
                stack = stack[:l-1]
            }else{
                return false
            }
        }else{
            stack = append(stack,char)
        }
    }
    
    return len(stack) == 0
}
